### DISCUSSION TITLE
DriverConstants.h Question: es_output_device_t

### DISCUSSION THREAD

**Discussion by ki6syd** [pschatzmann/arduino-audio-driver#21] [Q&A] [answered]
I am bringing up a custom board based on ESP32-S3 and ES8388, and ran into confusing hardware behavior when enabling only one of the DACs inside the ES8388. The codec outputs a clean signal when I enable all DAC outputs, but I'm finding that the wrong output is disabled when I try to power down an output. 

The definition of the `es_output_device_t` enum does not match what I would expect when reading the ES8388 datasheet. An example: [DriverConstants.h](https://github.com/pschatzmann/arduino-audio-driver/blob/8e88dc186d74dec0af4a9ccc1c07b308076fc737/src/Driver/DriverConstants.h#L15) defines `_DAC_OUTPUT_LOUT1 = 0x04`. From section 6.1.5 of the datasheet, I would expect 0x04 to correspond to ROUT2. Note that bits 0 and 1 are unused in this register.

![image](https://github.com/pschatzmann/arduino-audio-driver/assets/5254153/d043fd06-ef74-4e6f-a918-f7b6a4e755df)

[This](https://www.lcsc.com/datasheet/lcsc_datasheet_1912111437_Everest-semi-Everest-Semiconductor-ES8388_C365736.pdf) is the datasheet that I am using for ES8388.

Anyone able to explain where I'm going wrong in checking the datasheet against the enum definition? Thank you! 

Max

P.S. - this is my first time posting a discussion on GitHub, any feedback on how to ask good questions is appreciated.

**Comment by pschatzmann** [comment, marked as answer]
I was just taking the code that was provided by the Espressif ADF framework.
As far as I can see, the es8388_config_output_device() method is not used anywhere in my framework and the devices are defined via es8388_init()

If you want to correct this method, I suggest that you add a switch/case logic to the es8388.c. I am always happy to get  corrections via a pull request.

ps. how about getting rid of the enum completly and use int values instead of es_output_device or es_input_device ?

**Reply by ki6syd** [reply]
Thank you for the quick reply! 

That makes sense. I'll check for any evidence that others have found something similar in the Espressif ADF, and then test out a correction. I was disabling the output via es8388_init(). 

For the switch/case logic, do you mean this for both es8388_init() and es8388_config_output_device()? I can't say I understand the use case for es8388_config_output_device(), which looks like it can only enable a DAC. Avoiding the enum makes sense to me - there are already writes to ES8388_DACPOWER that aren't based on es_output_device.

**Reply by ki6syd** [reply]
@pschatzmann , following up on this one: I couldn't find anybody calling out an error in Espressif's code, but I did find [this driver](https://github.com/thaaraak/ESP32-ES8388/blob/master/main/codec_es8388.h) with different values that match the datasheet. I do think there is an error and `es_output_device_t` should be:

```
typedef enum {
    DAC_OUTPUT_MIN = -1,
    DAC_OUTPUT_LOUT1 = 0x20,
    DAC_OUTPUT_LOUT2 = 0x08,
    DAC_OUTPUT_SPK   = 0x09,
    DAC_OUTPUT_ROUT1 = 0x10,
    DAC_OUTPUT_ROUT2 = 0x04,
    DAC_OUTPUT_ALL = 0x3c,
    DAC_OUTPUT_MAX,
} es_dac_output_t;
```

Let me know if it'd be helpful for me to test the change on my hardware and submit a PR. Wasn't sure how you felt about diverging from Espressif's driver.

**Comment by pschatzmann** [comment]
As far as i know es_output_device_t does not use es_output_device_t, so there shouldn't be anything to correct there.
What exactly is your use case that you want to implement ?

The idea of this library is to use the [AudioBoard](https://pschatzmann.github.io/arduino-audio-driver/html/classaudio__driver_1_1_audio_board.html) methods and not the low lever drivers

**Reply by ki6syd** [reply]
I need to turn specific DACs on/off on-the-fly in my application. The hardware is a radio transceiver where a single ES8388 is sitting between the ESP32, headphones, mic, and audio in/out from multiple RF modulator/demodulators. I'm planning to turn one of the DACs off during radio RX mode, but was just doing the initial hardware bringup when I found the surprising behavior.

I love the library concept and am planning to work through the AudioBoard methods! I just ended up looking in the es8388 driver to debug this.

**Comment by pschatzmann** [comment]
To support TDS is started to add the method setMute(bool, int line) to some Drivers
I added now a DRAFT implementation to the [AudioDriverES8388Class](https://pschatzmann.github.io/arduino-audio-driver/html/classaudio__driver_1_1_audio_driver_e_s8388_class.html) as well

**Reply by ki6syd** [reply]
Thanks, Phil. I pulled your changes and am doing something like `audio_board.getDriver()->setMute(false, 0)` to test.

I'm trying this without success and think maybe line 794/[796](https://github.com/pschatzmann/arduino-audio-driver/blob/6d1571981a545404b561bdef4fdc67d3f7ce4a45/src/Driver.h#L796) should be swapped? Looks like the call to `setMute(false)` is undoing `line_active[line] = !mute` right now.

**Comment by pschatzmann** [comment]
I committed a correction. Now I using the right es_dac_output_t values and the output device with setMute(mute, line) is determined correctly as well. 

But somehow I do not manage to mute the speaker on my AudioKit and I don't manage to spot the error.
Maybe you can see what I am missing...

ps: your proposed enum is also mixing things up and does not correspond to the datasheet.

**Reply by ki6syd** [reply]
Running your latest commit on my board, it's properly muting/unmuting everything for me. 

Maybe try alternating between muted/unmuted and see if there is a volume difference? I am finding that when I don't load the ES8388 output with a low impedance, it still has some output signal present even when muted. I'd guess the speaker amp on the AudioKit board has a high impedance input and is still amplifying something.

Sorry about that, I did give that enum the wrong name in the previous comment - It should be es_**output_device**_t and have the ES8388_ prefix. I don't understand the need for `_MIN`, `_SPK`, `_MAX` so I left those alone. What about the datasheet do you see as an issue for LOUT1, LOUT2, ROUT1, ROUT2? Bits 0/1 of the register are unused in the datasheet I linked on the original post.


```
typedef enum {
  ES8388_OUTPUT_MIN = -1,
  ES8388_OUTPUT_LOUT1 = 0x20,
  ES8388_OUTPUT_ROUT1 = 0x08,
  ES8388_OUTPUT_SPK   = 0x09,
  ES8388_OUTPUT_LOUT2 = 0x10,
  ES8388_OUTPUT_ROUT2 = 0x04,
  ES8388_OUTPUT_ALL = 0x3c,
  ES8388_OUTPUT_MAX,
} es_output_device_t;
```

**Comment by pschatzmann** [comment]
In my latest commit, I made sure, that the line number matches with the board definition: LINE1 goes now to LOUT1 and ROUT1.
When the line 0 is muted, I am now also switching off the PA, to make the setMute(mute, line); work on the AudioKit board!

This behavior can be configured in the new AudioDriverConfig.h
